capstone/src/main/java/com/google/sps/AssociationDatastore.java

```diff
@@ -0,0 +1,104 @@
+package com.google.sps;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.sps.data.MapData;
+import java.util.ArrayList;
+import java.util.Optional;
+
+/** Does all datastore operations related to associations */
+public class AssociationDatastore {
+
+  public static final String SURVEY_ENTITY_KIND = "Response";
+  public static final String COMMENT_PROPERTY = "text";
+  public static final String ZIPCODE = "zipCode";
+
+  private MapData mapData = new MapData();
+  private DatastoreService datastore;
+
+  /** @param datastore the datastore client */
+  public AssociationDatastore(DatastoreService datastore) {
+    this.datastore = datastore;
+  }
+
+  /**
+   * Get the survey comment text from datastore
+   *
+   * @return an arraylist of the comment text
+   */
+  public ArrayList<AssociationInput> getComments() {
+    Query query = new Query(SURVEY_ENTITY_KIND);
+    PreparedQuery results = datastore.prepare(query);
+
+    ArrayList<AssociationInput> comments = new ArrayList<AssociationInput>();
+    for (Entity e : results.asIterable()) {
+      if (e.getProperty("association-processed") == null
+          || !((boolean) e.getProperty("association-processed"))) {
+        String message = (String) e.getProperty(COMMENT_PROPERTY);
+        try {
+          ArrayList<String> scope = mapData.getScope((String) e.getProperty(ZIPCODE));
+          comments.add(new AssociationInput(message, scope));
+          e.setProperty("association-processed", true);
+          datastore.put(e);
+        } catch (NullPointerException exception) {
```

Comment:
Generally `NullPointerException`s like this are systemic of an underlying bug and can catch exceptions that we did not intend to (due to bugs). Can we do a check before this intended `NullPointerException` happens?

Reply:
I believe that the original source of this was that there were some leftover synthetic data results mixed in with actual responses which generated null pointers off of undefined properties. I can remove the try catch and instead check for null properties when I access them.

Reply:
Thanks, that sounds like a good way to approach it.